pkg/build/controller/common/util.go

```diff
@@ -42,6 +47,75 @@ func HandleBuildCompletion(build *buildapi.Build, runPolicies []policy.RunPolicy
 	if err := runPolicy.OnComplete(build); err != nil {
 		utilruntime.HandleError(fmt.Errorf("failed to run policy on completed build %s/%s: %v", build.Namespace, build.Name, err))
 	}
+	if err := HandleBuildPruning(buildutil.ConfigNameForBuild(build), build.Namespace, buildLister, buildConfigGetter, buildDeleter); err != nil {
+		utilruntime.HandleError(fmt.Errorf("failed to prune old builds %s/%s: %v", build.Namespace, build.Name, err))
+	}
+}
+
+type ByCreationTimestamp []buildapi.Build
+
+func (b ByCreationTimestamp) Len() int {
+	return len(b)
+}
+
+func (b ByCreationTimestamp) Swap(i, j int) {
+	b[i], b[j] = b[j], b[i]
+}
+
+func (b ByCreationTimestamp) Less(i, j int) bool {
+	return !b[j].CreationTimestamp.Time.After(b[i].CreationTimestamp.Time)
+}
+
+// HandleBuildPruning handles the deletion of old successful and failed builds
+// based on settings in the BuildConfig.
+func HandleBuildPruning(buildConfigName string, namespace string, buildLister buildclient.BuildLister, buildConfigGetter buildclient.BuildConfigGetter, buildDeleter buildclient.BuildDeleter) error {
```

Comment:
i'm still a little bothered this function isn't using the shared informer caches to get the buildconfig+builds, but we can fix it in a follow up.  Please open an issue to track the need to deal with it.